dot/parachain/dispute/participation.go

```diff
@@ -0,0 +1,376 @@
+
+func (p *ParticipationHandler) Queue(ctx overseer.Context,
+	request ParticipationRequest,
+	priority ParticipationPriority,
+) error {
+	if _, ok := p.runningParticipation.Load(request.candidateHash); ok {
+		return nil
+	}
+
+	// if we already have a recent block, participate right away
+	if p.recentBlock != nil && p.numberOfWorkers() < MaxParallelParticipation {
+		p.forkParticipation(&request, p.recentBlock.Hash)
+		return nil
+	}
+
+	blockNumber, err := getBlockNumber(ctx.Sender, request.candidateReceipt)
+	if err != nil {
+		return fmt.Errorf("get block number: %w", err)
+	}
+
+	comparator, err := NewCandidateComparator(&blockNumber, request.candidateReceipt)
+	if err != nil {
+		return fmt.Errorf("create candidate comparator: %w", err)
+	}
+
+	if err := p.queue.Queue(comparator, &request, priority); err != nil {
+		return fmt.Errorf("queue ParticipationHandler request: %w", err)
+	}
+
+	return nil
+}
+
+func (p *ParticipationHandler) Clear(candidateHash common.Hash) error {
+	p.runningParticipation.Delete(candidateHash)
+	p.workers.Add(-1)
+
+	if p.recentBlock == nil {
+		panic("we never ever reset recentBlock to nil and we already received a result, so it must have been set before. qed")
+	}
+
+	p.dequeueUntilCapacity(p.recentBlock.Hash)
+	return nil
+}
+
+func (p *ParticipationHandler) ProcessActiveLeavesUpdate(update overseer.ActiveLeavesUpdate) error {
+	if update.Activated == nil {
+		return nil
+	}
+
+	if p.recentBlock == nil {
+		p.recentBlock = &block{
+			Number: update.Activated.Number,
+			Hash:   update.Activated.Hash,
+		}
+
+		p.dequeueUntilCapacity(update.Activated.Hash)
+	} else {
+		if update.Activated.Number > p.recentBlock.Number {
+			p.recentBlock.Number = update.Activated.Number
+			p.recentBlock.Hash = update.Activated.Hash
+		}
+	}
+
+	return nil
+}
+
+func (p *ParticipationHandler) BumpPriority(ctx overseer.Context, receipts []parachainTypes.CandidateReceipt) error {
+	for _, receipt := range receipts {
+		blockNumber, err := getBlockNumber(ctx.Sender, receipt)
+		if err != nil {
+			logger.Errorf(
+				"failed to get block number. CommitmentsHash: %s, Error: %s",
+				receipt.CommitmentsHash.String(),
+				err,
+			)
+			continue
+		}
+
+		comparator, err := NewCandidateComparator(&blockNumber, receipt)
+		if err != nil {
+			logger.Errorf(
+				"failed to create candidate comparator. CommitmentsHash: %s, Error: %s",
+				receipt.CommitmentsHash.String(),
+				err,
+			)
+			continue
+		}
+
+		if err := p.queue.PrioritiseIfPresent(comparator); err != nil {
+			logger.Errorf(
+				"failed to prioritise candidate. CommitmentsHash: %s, Error: %s",
+				receipt.CommitmentsHash.String(),
+				err,
+			)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (p *ParticipationHandler) numberOfWorkers() int {
+	return int(p.workers.Load())
+}
+
+func (p *ParticipationHandler) dequeueUntilCapacity(recentHead common.Hash) {
+	for {
+		if p.numberOfWorkers() >= MaxParallelParticipation {
+			break
+		}
+
```

Comment:
I think you can write this in a single line
```suggestion
	for p.numberOfWorkers() < MaxParallelParticipation {
```

Reply:
I agree, that's neater. Thank you!